locallib.php

```diff
@@ -443,21 +443,51 @@ function local_boostnavigation_cohort_is_member($userid, $setting) {
     // Initialize variable for memberships.
     static $allmemberships = null;
 
-    // First: If the memberships haven't been fetched yet initially, fetch all of the user's cohort memberships
-    // only once and remember for next calls of this function.
-    if ($allmemberships == null) {
-        // Prepare SQL statement.
-        $sql = 'SELECT {cohort}.id, {cohort}.idnumber FROM {cohort_members}
-            JOIN {cohort}
-                ON {cohort}.id = {cohort_members}.cohortid
-            WHERE {cohort_members}.userid = ?';
-        $params = array($userid);
-
-        // Run DB query.
-        $ret = $DB->get_records_sql_menu($sql, $params);
-
-        // Remember memberships.
-        $allmemberships = $ret;
+    try {
+        $cohortversioncache = cache::make('local_boostnavigation', 'cohortversion');
+        $usercohortversioncache = cache::make('local_boostnavigation', 'usercohortversion');
+        $usercohortscache = cache::make('local_boostnavigation', 'usercohorts');
+    } catch (Exception $e) {
+        throw $e;
+    }
+
+    // Set cohort version cache with current timestamp if the cohort version is not set
+    // from the cohort event observers.
+    if (!$cohortversioncache->get('cohortversion')) {
+        $cohortversioncache->set('cohortversion', time());
+    }
+
+    // Check if the user cohort version and cohort version cache are not same, then we need to read cohorts from db
+    if (!$usercohortscache->get('usercohorts') ||
+        !$usercohortversioncache->get('usercohortversion') ||
+        $cohortversioncache->get('cohortversion') != $usercohortversioncache->get('usercohortversion')) {
+
+            // Check if the cohort version is future time(debounce time of 1 minute is set in the cohort event observers)
+            // then read cohorts from cache.
+            if ($cohortversioncache->get('cohortversion') > time() && $usercohortscache->get('usercohorts')) {
```

Comment:
Another small comment, avoid getting the same cache item again as you might be making a second call to redis etc.  You can statically cache this inside the cache definition so core will do this for you, but it's still a good habit to only call the get onces:

'staticacceleration' => false,

Reply:
Thank you Brendan !!

I've fixed this as mentioned. Please review.

Thanks,
Anu